onnxruntime/core/providers/webnn/builders/impl/slice_op_builder.cc

```diff
@@ -89,31 +82,59 @@ Status SliceOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
   ORT_RETURN_IF_ERROR(
       SliceOp::PrepareForComputeHelper(input_starts, input_ends, input_axes, input_steps, compute_metadata));
 
-  std::transform(compute_metadata.starts_.cbegin(), compute_metadata.starts_.cend(),
-                 starts.begin(),
-                 [](int64_t i) { return SafeInt<uint32_t>(i); });
-  std::transform(compute_metadata.ends_.cbegin(), compute_metadata.ends_.cend(), compute_metadata.starts_.cbegin(),
-                 sizes.begin(),
-                 [](int64_t i, int64_t j) { return SafeInt<uint32_t>(i - j); });
-  std::transform(compute_metadata.steps_.cbegin(), compute_metadata.steps_.cend(), steps.begin(),
-                 [](int64_t i) { return SafeInt<uint32_t>(i); });
-
-  emscripten::val options = emscripten::val::object();
-  options.set("strides", emscripten::val::array(steps));
-  options.set("label", node.Name());
-  emscripten::val output = model_builder.GetBuilder().call<emscripten::val>("slice", inputs,
-                                                                            emscripten::val::array(starts),
-                                                                            emscripten::val::array(sizes),
-                                                                            options);
+  // Check if reverse op is needed.
+  std::vector<uint32_t> reverse_axes;
+  emscripten::val reverse_output = input;
+  for (size_t i = 0; i < rank; ++i) {
+    if (compute_metadata.steps_[i] < 0) {
+      reverse_axes.push_back(SafeInt<uint32_t>(i));
+      compute_metadata.steps_[i] = -compute_metadata.steps_[i];
+      compute_metadata.starts_[i] = input_shape[i] - 1 - compute_metadata.starts_[i];
+      compute_metadata.ends_[i] = input_shape[i] - 1 - compute_metadata.ends_[i];
+    }
+  }
+  if (!reverse_axes.empty()) {
+    emscripten::val reverse_options = emscripten::val::object();
+    reverse_options.set("axes", emscripten::val::array(reverse_axes));
+    reverse_options.set("label", node.Name() + "_reverse");
+    reverse_output = model_builder.GetBuilder().call<emscripten::val>("reverse", input, reverse_options);
+  }
+
+  // Check if slice op is needed.
+  bool is_slice_required = false;
+  for (size_t i = 0; i < rank; ++i) {
+    if (compute_metadata.steps_[i] != 1 || compute_metadata.starts_[i] != 0 ||
+        compute_metadata.ends_[i] != input_shape[i]) {
+      is_slice_required = true;
+      break;
+    }
+  }
+
+  emscripten::val output = reverse_output;
+  if (is_slice_required) {
+    std::vector<uint32_t> starts(rank);
+    std::vector<uint32_t> sizes(rank);
+    std::vector<uint32_t> steps(rank);
+    std::transform(compute_metadata.starts_.cbegin(), compute_metadata.starts_.cend(), starts.begin(),
```

Comment:
Hmm, isn't there already a helper function somewhere under the WebNN EP to convert to uint32? We see this pattern so often.

```
std::vector<uint32_t> steps = ConvertArrayToUint32(compute_metadata.steps_);
```
(obviously the sizes via i - j still needs its own transform call)

Reply:
We have `GetVecUint32FromVecInt64` function in `helper.h`.

Reply:
> Hmm, isn't there already a helper function somewhere under the WebNN EP to convert to uint32? We see this pattern so often.

I didn't find this kind of helper function, @Honry do you know?

Reply:
Done, thanks!